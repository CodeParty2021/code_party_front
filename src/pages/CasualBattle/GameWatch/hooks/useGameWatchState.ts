import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import {
  ResultType,
  useFetchResult,
} from "hooks/ResultAPIHooks/useFetchResult";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";

export type IResponse = {
  isAnalyzing: boolean;
  analyzingError: boolean;
  result?: ResultType;
  json?: string;
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: result, fetchResult } = useFetchResult();
  const { data: json, fetchJson } = useFetchJson();
  const navigate = useNavigate();

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/casual-battle");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchResult(room.info?.analyzingResult.resultId);
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  const _exitBtnHandler = () => {
    navigate("/casual-battle/waiting-room");
  };

  return {
    isAnalyzing: !room.info?.analyzingResult?.resultId,
    analyzingError: Boolean(room.info?.analyzingResult?.error),
    result: result,
    json: json,
    exitBtnHandler: _exitBtnHandler,
  };
};
