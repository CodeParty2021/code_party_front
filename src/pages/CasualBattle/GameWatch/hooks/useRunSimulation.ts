import { useEffect } from "react";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useRunCodes } from "hooks/CodeAPIHooks/useRunCodes";

export type IResponse = {};

export const useRunSimulation = (): IResponse => {
  const { room, isHost, updateRoomInfo } = useRoomSync();
  const {
    data: simulationResult,
    error: simulationError,
    runCodes,
  } = useRunCodes();

  //初期処理
  useEffect(() => {
    //シミュレーションの実行(ホストのみ実行)
    if (isHost) {
      const codes: string[] = [];
      room.sortedKeysOfMembers.forEach((key) => {
        const member = room.members[key];
        if (member.codeId) codes.push(member.codeId);
      });
      runCodes(...codes);
    }
  }, []);

  //シミュレーション結果を受信した場合
  useEffect(() => {
    if (simulationResult) {
      updateRoomInfo({
        analyzingResult: {
          resultId: simulationResult.jsonId,
        },
      });
    }
  }, [simulationResult]);

  //シミュレーション中にエラーが発生した場合
  useEffect(() => {
    if (simulationError) {
      updateRoomInfo({
        analyzingResult: {
          error: "シミュレーション実行中にエラーが発生しました。",
        },
      });
    }
  }, [simulationError]);

  return {};
};
