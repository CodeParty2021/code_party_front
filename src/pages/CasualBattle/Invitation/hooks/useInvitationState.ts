import { useEffect, useState } from "react";
import { Params, useNavigate, useParams } from "react-router-dom";
import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";

export type IResponse = {
  isLogin: boolean;
  errorMessage?: string;
  roomId?: string;
  enterBtnDisabled: boolean;
  enterBtnClickHandler: () => void;
};

export const useInvitationState = (): IResponse => {
  const { roomId } = useParams<Params<string>>();
  const { isLogin } = useSelector((state: RootState) => state.user);
  const [enterBtnDisabled, setEnterBtnDisabled] = useState(false);
  const { room, enterRoom } = useRoomSync();
  const [errorMessage, setErrorMessage] = useState<string | undefined>(
    undefined
  );
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/casual-battle/waiting-room");
    }
  }, [room.isEntered]);

  //入場ボタン
  const _enterBtnHandler = () => {
    if (typeof roomId === "string") {
      setEnterBtnDisabled(true);
      enterRoom(roomId).catch(() => {
        setEnterBtnDisabled(false);
      });
    } else {
      setErrorMessage("招待リンクが無効です。");
    }
  };

  return {
    isLogin: isLogin,
    errorMessage: errorMessage,
    roomId: roomId,
    enterBtnDisabled: enterBtnDisabled,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
