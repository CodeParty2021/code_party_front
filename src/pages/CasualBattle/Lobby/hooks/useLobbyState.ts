import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  roomCreateBtnHandler: () => void;
  roomSearchBtnHandler: () => void;
};

export const useLobbyState = (): IResponse => {
  const { room, createRoom } = useRoomSync();
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/casual-battle/waiting-room");
    }
  }, [room.isEntered]);

  const _roomCreateBtnHandler = () => {
    createRoom();
  };

  const _roomSearchBtnHandler = () => {
    navigate("/casual-battle/search-room");
  };

  return {
    roomCreateBtnHandler: _roomCreateBtnHandler,
    roomSearchBtnHandler: _roomSearchBtnHandler,
  };
};
