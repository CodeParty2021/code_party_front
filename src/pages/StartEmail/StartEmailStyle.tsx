import styled from "styled-components";
import { WHITE } from "styles/colors";
import { FONT } from "styles/constants/constants";

export const BackLinkPosition = styled.div`
  position: absolute;
  top: 32px;
  left: 32px;
`;

export const TextInput = styled.input`
  width: 453px;
  height: 64px;
  background: ${WHITE};
  border-radius: 8px;
  font-family: ${FONT.NOTO_SANS};
  font-style: normal;
  font-weight: 400;
  font-size: 24px;
  line-height: 150%;
  text-align: center;
  color: #2a2c33;
  margin-bottom: 32px;
`;
