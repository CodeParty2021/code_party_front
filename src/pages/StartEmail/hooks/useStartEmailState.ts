import { getAuth } from "firebase/auth";
import { useFirebaseAuth } from "hooks/FirebaseAuthHooks/useFirebaseAuthHooks";
import { RefObject, useEffect, useRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { signOutAsync } from "services/user/user";
import { RootState } from "store";

export type IResponse = {
  loading: boolean;
  algoMessage: string;
  emailRef: RefObject<HTMLInputElement>;
  passwordRef: RefObject<HTMLInputElement>;
  emailChangeHandler: () => void;
  passwordChangeHandler: () => void;
  startBtnHandler: () => void;
  btnDisabled: boolean;
  backLinkButtonHandler: () => void;
};

export const useStartState = (): IResponse => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const emailInputRef = useRef<HTMLInputElement>(null);
  const passwordInputRef = useRef<HTMLInputElement>(null);
  const [btnDisabled, setBtnDisabled] = useState<boolean>(true);
  const [loading, setLoading] = useState(true);
  const [algoMessage, setAlgoMessage] =
    useState("メールアドレスとパスワードをいれてね");

  const isLogin = useSelector((state: RootState) => state.user.isLogin);
  const auth = getAuth();

  const { signInWithEmail } = useFirebaseAuth();

  useEffect(() => {
    dispatch(signOutAsync());
    setLoading(false);
  }, []);

  const _updateBtnDisabledState = () => {
    if (
      emailInputRef.current?.value === "" ||
      passwordInputRef.current?.value === ""
    ) {
      setBtnDisabled(true);
    } else {
      setBtnDisabled(false);
    }
  };

  const backLinkButtonHandler = (): void => {
    navigate("/", { replace: true });
  };

  const startBtnHandler = async () => {
    setLoading(true);
    const email = emailInputRef.current?.value as string;
    const password = passwordInputRef.current?.value as string;
    console.log(email, password);
    await signInWithEmail({
      email,
      password,
    }).catch(() => {
      setAlgoMessage("メールアドレスかパスワードが違います");
      setLoading(false);
    });
    setLoading(false);
  };

  if (isLogin && auth.currentUser && !auth.currentUser?.isAnonymous) {
    navigate("/set-name", { replace: true });
  }

  return {
    loading,
    algoMessage,
    emailRef: emailInputRef,
    passwordRef: passwordInputRef,
    emailChangeHandler: _updateBtnDisabledState,
    passwordChangeHandler: _updateBtnDisabledState,
    startBtnHandler,
    backLinkButtonHandler,
    btnDisabled,
  };
};
