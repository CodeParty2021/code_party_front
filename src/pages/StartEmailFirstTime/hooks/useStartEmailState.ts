import { getAuth } from "firebase/auth";
import { useFirebaseAuth } from "hooks/FirebaseAuthHooks/useFirebaseAuthHooks";
import { RefObject, useEffect, useRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { signOutAsync } from "services/user/user";
import { RootState } from "store";

export type IResponse = {
  loading: boolean;
  algoMessage: string;
  emailRef: RefObject<HTMLInputElement>;
  passwordRef: RefObject<HTMLInputElement>;
  emailChangeHandler: () => void;
  passwordChangeHandler: () => void;
  startBtnHandler: () => void;
  btnDisabled: boolean;
  backLinkButtonHandler: () => void;
};

export const useStartEmailState = (): IResponse => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const emailInputRef = useRef<HTMLInputElement>(null);
  const passwordInputRef = useRef<HTMLInputElement>(null);
  const [btnDisabled, setBtnDisabled] = useState<boolean>(true);
  const [loading, setLoading] = useState(true);
  const [algoMessage, setAlgoMessage] =
    useState("メールアドレスとパスワードをいれてね");
  const { createUserWithEmail } = useFirebaseAuth();

  useEffect(() => {
    dispatch(signOutAsync());
    setLoading(false);
  }, []);

  const isLogin = useSelector((state: RootState) => state.user.isLogin);
  const auth = getAuth();

  const _updateBtnDisabledState = () => {
    if (
      emailInputRef.current?.value === "" ||
      passwordInputRef.current?.value === ""
    ) {
      setBtnDisabled(true);
    } else {
      setBtnDisabled(false);
    }
  };

  const backLinkButtonHandler = (): void => {
    navigate("/", { replace: true });
  };

  const startBtnHandler = async () => {
    setLoading(true);
    const email = emailInputRef.current?.value as string;
    const password = passwordInputRef.current?.value as string;
    console.log(email, password);
    await createUserWithEmail({
      email,
      password,
    }).catch(() => {
      setAlgoMessage("違うメールアドレスを試してみてください");
      setLoading(false);
    });
  };

  if (isLogin && auth.currentUser && !auth.currentUser?.isAnonymous) {
    navigate("/set-name", { replace: true });
  }
  return {
    loading,
    algoMessage,
    emailRef: emailInputRef,
    passwordRef: passwordInputRef,
    emailChangeHandler: _updateBtnDisabledState,
    passwordChangeHandler: _updateBtnDisabledState,
    startBtnHandler,
    backLinkButtonHandler,
    btnDisabled,
  };
};
