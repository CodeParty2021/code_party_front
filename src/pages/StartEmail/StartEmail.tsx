import { BackLink } from "components/BackLink/BackLink";
import { Button } from "components/Button/Button";
import { StarBackground } from "components/StarBackground/StarBackground";
import { TranslucentCard } from "components/TranslucentCard/TranslucentCard";
import { Loading } from "pages/Loading/Loading";
import React from "react";
import { useStartState } from "./hooks/useStartEmailState";
import { BackLinkPosition, TextInput } from "./StartEmailStyle";

export const StartEmail: React.FC = () => {
  const {
    loading,
    algoMessage,
    emailRef,
    passwordRef,
    emailChangeHandler,
    passwordChangeHandler,
    startBtnHandler,
    backLinkButtonHandler,
    btnDisabled,
    enterSubmitHandler,
  } = useStartState();
  return loading ? (
    <Loading />
  ) : (
    <StarBackground>
      <BackLinkPosition>
        <BackLink
          backMessage="トップに戻る"
          onClick={backLinkButtonHandler}
          iconColor="black"
        />
      </BackLinkPosition>
      <TranslucentCard modalTitle="ログインして遊ぶ" algoMessage={algoMessage}>
        <TextInput
          ref={emailRef}
          type="email"
          onChange={emailChangeHandler}
          onKeyDown={enterSubmitHandler}
          placeholder="メールアドレス"
        ></TextInput>
        <TextInput
          ref={passwordRef}
          type="password"
          onChange={passwordChangeHandler}
          onKeyDown={enterSubmitHandler}
          placeholder="パスワード"
        ></TextInput>
        <Button
          onClick={startBtnHandler}
          color="pink"
          size="M"
          value="ログイン"
          status={btnDisabled ? "disabled" : "default"}
        />
      </TranslucentCard>
    </StarBackground>
  );
};
