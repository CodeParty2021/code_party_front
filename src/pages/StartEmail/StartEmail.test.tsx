import React from "react";
import { IResponse, useStartState } from "./hooks/useStartEmailState";
import { shallow } from "enzyme";
import { StartEmail } from "./StartEmail";

jest.mock("./hooks/useStartEmailState");

const useStartStateMock = useStartState as jest.Mock;

const initialState: IResponse = {
  loading: false,
  algoMessage: "ログインしてください",
  emailRef: { current: null },
  passwordRef: { current: null },
  emailChangeHandler: jest.fn(),
  passwordChangeHandler: jest.fn(),
  startBtnHandler: jest.fn(),
  btnDisabled: false,
  backLinkButtonHandler: jest.fn(),
};

describe("<StartEmailFirstTime />", () => {
  beforeEach(() => {
    useStartStateMock.mockReturnValue({ ...initialState });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });
  it("auth snapshot test", () => {
    const wrapper = shallow(<StartEmail />);
    expect(wrapper.getElements()).toMatchSnapshot();
  });
});
