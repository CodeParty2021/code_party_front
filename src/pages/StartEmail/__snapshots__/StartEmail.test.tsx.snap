// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StartEmail /> auth snapshot test 1`] = `
Array [
  <StarBackground>
    <styled.div>
      <BackLink
        backMessage="トップに戻る"
        iconColor="black"
        onClick={[MockFunction]}
      />
    </styled.div>
    <TranslucentCard
      algoMessage="ログインしてください"
      modalTitle="ログインして遊ぶ"
    >
      <styled.input
        onChange={[MockFunction]}
        placeholder="メールアドレス"
        type="email"
      />
      <styled.input
        onChange={[MockFunction]}
        placeholder="パスワード"
        type="password"
      />
      <Button
        color="pink"
        onClick={[MockFunction]}
        size="M"
        status="default"
        value="ログイン"
      />
    </TranslucentCard>
  </StarBackground>,
]
`;
